plugins/statuslines/lualine.nix

```diff
@@ -1,177 +1,252 @@
-              default = { };
-              description = "extra options for the component";
-            };
-          };
-        })
-      )
-    ) "";
-
-  mkEmptySectionOption = name: {
-    lualine_a = mkComponentOptions "";
-    lualine_b = mkComponentOptions "";
-    lualine_c = mkComponentOptions "";
-    lualine_x = mkComponentOptions "";
-    lualine_y = mkComponentOptions "";
-    lualine_z = mkComponentOptions "";
-  };
-in
-{
-  options = {
-    plugins.lualine = {
-      enable = mkEnableOption "lualine";
-
-      package = lib.mkPackageOption pkgs "lualine" {
-        default = [
-          "vimPlugins"
-          "lualine-nvim"
-        ];
-      };
+              };
+            in
+            maybeRaw (listOf (either str (lib.nixvim.transitionType oldType coerceFn (maybeRaw newType))));
+          inherit description;
+        };
 
-      gitPackage = lib.mkPackageOption pkgs "git" {
-        nullable = true;
+      mkEmptySectionOption = name: {
+        lualine_a = mkComponentOptions "Left section on left side.";
+        lualine_b = mkComponentOptions "Middle section on left side.";
+        lualine_c = mkComponentOptions "Right section on left side.";
+        lualine_x = mkComponentOptions "Left section on right side.";
+        lualine_y = mkComponentOptions "Middle section on right side.";
+        lualine_z = mkComponentOptions "Right section on right side.";
       };
 
-      iconsEnabled = mkOption {
-        type = types.bool;
-        description = "Whether to enable/disable icons for all components.";
-        default = true;
-      };
+    in
+    {
+      options = {
+        icons_enabled = defaultNullOpts.mkBool true ''
+          Whether to enable icons for all components.
+        '';
 
-      theme = helpers.defaultNullOpts.mkNullable (
-        with types; either str attrs
-      ) "auto" "The theme to use for lualine-nvim.";
+        theme = defaultNullOpts.mkNullable (
+          with types; either str attrs
+        ) "auto" "The theme to use for lualine-nvim.";
 
-      componentSeparators = mkSeparatorsOption {
-        leftDefault = "";
-        rightDefault = "";
-      };
+        component_separators = mkSeparatorsOption {
+          left = "";
+          right = "";
+        };
 
-      sectionSeparators = mkSeparatorsOption {
-        leftDefault = "";
-        rightDefault = "";
-      };
+        section_separators = mkSeparatorsOption {
+          left = "";
+          right = "";
+        };
 
-      disabledFiletypes = {
-        statusline = helpers.defaultNullOpts.mkListOf types.str [ ] ''
-          Only ignores the ft for statusline.
-        '';
+        disabled_filetypes = defaultNullOpts.mkNullableWithRaw' {
+          description = ''
+            Filetypes in which to disable lualine.
+            Allows you to specify filetypes that you want to only disable on specific components.
+          '';
+          pluginDefault = { };
+          type = types.submodule {
```

Comment:
Awesome. Should we consider allowing list definitions? Either just a plain `either (listOf str) (submodule)` or a coerceTo using `lib.nixvim.mkUnkeyedAttrs`?

Both approaches have issues, for example using coercion here would not merge multiple list defs well, since they'd each end up with conflicting indexes.

Reply:
A propper `types.luaTable` would help, but that's _way_ out of scope for the PR :grin:

Reply:
I like the idea of allowing the list. Would allow it to be easier to use when you don't care about granular control.